### PR TITLE
Mark CTFE bitmanip function for generating longs to a string as @safe and pure

### PR DESCRIPTION
**Pull request by schveiguy**
Helps with proposed DIP1028 (https://github.com/dlang/dmd/pull/10709, https://github.com/dlang/DIPs/blob/1b705f8d4faa095d6d9e3a1b81d6cfa6d688554b/DIPs/DIP1028.md)

Not sure if there will be more. It's difficult to test this, but this seems to be the most common problem.

cc @WebFreak001

### CODE REVIEW

**Comment by WebFreak001** [comment]
ok fixed the bitfields issues in dub packages and has improved the quota so far a bit. What I see now is left when randomly sampling error logs:

- **`std.regex : ctRegex` doesn't compile**
- **`std.functional : toDelegate` broken** (actually nearly fixed by this PR aswell!)
- delegate arguments are still `@system` by default (toString with delegate sink)
- lots of libraries using `__gshared`
- usage of `.ptr`
- usage of `void*`
- cast from `void[]` to `ubyte[]` is not allowed
- calls of `core.stdc.stdlib` not `@safe`
- multiple overloads 

See https://wfr.moe/safe-test/phobos-7343.html (currently building)

But this PR is good to fix the bitfields issue.

**Comment by schveiguy** [comment]
OK good. Those sound more like real issues to be looked at.

I'll see if I can pick some more low hanging fruit...

**Comment by schveiguy** [comment]
> * delegate arguments are still `@system` by default (toString with delegate sink)

This should be fixed in the DIP PR, I would think. I can't imagine why they wouldn't be safe by default.

**Comment by WebFreak001** [comment]
@schveiguy I made a comment about it what it breaks

**Comment by WebFreak001** [comment]
there are also issues with `assumeUnique` and `assumeSafeAppend` being `@system`.

I can see why they are `@system` but one could argue they should be `@trusted` because they are simply just wrappers to make something unsafe safe

**Comment by aG0aep6G** [comment]
> there are also issues with `assumeUnique` and `assumeSafeAppend` being `@system`.
> 
> I can see why they are `@system` but one could argue they should be `@trusted` because they are simply just wrappers to make something unsafe safe

No. `assumeUnique` and `assumeSafeAppend` don't make anything safe.

**Comment by WebFreak001** [comment]
ok dub packages are all tested. We went from
```
compiles: 316, broken: 902, n/a: 499
works: 25.9%
```
to
```
compiles: 448, broken: 770, n/a: 499
works: 36.8%
```
+9.9% improvement in all of dub!

File: std/bitmanip.d [hunk 1]
```diff
@@ -56,14 +56,26 @@ import std.range.primitives;
...
 import std.traits;
 
-private string myToString(ulong n)
+private string myToString(ulong n) pure @safe
 {
-    import core.internal.string : UnsignedStringBuf, unsignedToTempString;
-    UnsignedStringBuf buf;
-    auto s = unsignedToTempString(n, buf);
-    return cast(string) s ~ (n > uint.max ? "UL" : "U");
+    static char[] generateStr(ulong n) pure
```
Comment (aG0aep6G) [inline std/bitmanip.d]: Do you really need the second function? This seems to work for me:

```d
string myToString(ulong n) pure @safe
{
    import core.internal.string : UnsignedStringBuf, unsignedToTempString;
    UnsignedStringBuf buf;
    auto s = unsignedToTempString(n, buf);
    return s ~ (n > uint.max ? "UL" : "U");
}
```

Comment (schveiguy) [reply std/bitmanip.d]: first thing I tried was returning char[] and then I thought I don’t need to for this case and I didn’t try your idea. But not really any reason except I thought that wouldn’t work. I’ll switch to your idea

Comment (schveiguy) [reply std/bitmanip.d]: OK, I updated to just remove the cast, and add pure/safe.

File: std/bitmanip.d [hunk 2]
```diff
@@ -56,14 +56,22 @@ import std.range.primitives;
...
 import std.traits;
 
-private string myToString(ulong n)
+private string myToString(ulong n) pure @safe
 {
     import core.internal.string : UnsignedStringBuf, unsignedToTempString;
     UnsignedStringBuf buf;
     auto s = unsignedToTempString(n, buf);
-    return cast(string) s ~ (n > uint.max ? "UL" : "U");
+    return s ~ (n > uint.max ? "UL" : "U");
```
Comment (WebFreak001) [inline std/bitmanip.d]: I think with the comment "use pure to implicitly cast to string" this was much better

Comment (schveiguy) [reply std/bitmanip.d]: Fixed.

**Comment by schveiguy** [comment]
Thanks. Is there a reason why auto merge label didn't work here? Is it the buildkite failure? @wilzbach

**Pull request merged**